File: app/src/main/java/io/nekohasekai/sagernet/fmt/tuic/TuicBean.java
package io.nekohasekai.sagernet.fmt.tuic;

import androidx.annotation.NonNull;
import com.esotericsoftware.kryo.io.ByteBufferInput;
import com.esotericsoftware.kryo.io.ByteBufferOutput;
import io.nekohasekai.sagernet.fmt.AbstractBean;
import io.nekohasekai.sagernet.fmt.KryoConverters;
import io.nekohasekai.sagernet.fmt.SingBoxOptions;

import org.jetbrains.annotations.NotNull;

public class TuicBean extends AbstractBean {

    public static final Creator<TuicBean> CREATOR = new CREATOR<TuicBean>() {
        @NonNull
        @Override
        public TuicBean newInstance() {
            return new TuicBean();
        }

        @Override
        public TuicBean[] newArray(int size) {
            return new TuicBean[size];
        }
    };
    public String token;
    public String certificates;
    public String certPublicKeySha256;
    public String udpRelayMode;
    public String congestionController;
    public String alpn;
    public Boolean disableSNI;
    public Boolean zeroRTT;
    public Integer mtu;

    // TUIC zep
    public String sni;

    public Boolean allowInsecure;
    public String customJSON;
    public String uuid;

    // ECH
    public Boolean ech;
    public String echConfig;
    public String echQueryServerName;

    // mTLS
    public String clientCert;
    public String clientKey;

    @Override
    public void initializeDefaultValues() {
        super.initializeDefaultValues();
        if (token == null) token = "";
        if (certificates == null) certificates = "";
        if (certPublicKeySha256 == null) certPublicKeySha256 = "";
        if (udpRelayMode == null) udpRelayMode = "native";
        if (congestionController == null) congestionController = "cubic";
        if (alpn == null) alpn = "";
        if (disableSNI == null) disableSNI = false;
        if (zeroRTT == null) zeroRTT = false;
        if (mtu == null) mtu = 1400;
        if (sni == null) sni = "";
        if (allowInsecure == null) allowInsecure = false;
        if (customJSON == null) customJSON = "";
        if (uuid == null) uuid = "";
        if (ech == null) ech = false;
        if (echConfig == null) echConfig = "";
        if (echQueryServerName == null) echQueryServerName = "";
        if (clientCert == null) clientCert = "";
        if (clientKey == null) clientKey = "";
    }

    @Override
    public void serialize(ByteBufferOutput output) {
        output.writeInt(3);

        // version 0
        super.serialize(output);
        output.writeString(token);
        output.writeString(certificates);
        output.writeString(udpRelayMode);
        output.writeString(congestionController);
        output.writeString(alpn);
        output.writeBoolean(disableSNI);
        output.writeBoolean(zeroRTT);
        output.writeInt(mtu);
        output.writeString(sni);
        output.writeBoolean(allowInsecure);
        output.writeString(customJSON);
        output.writeString(uuid);
        output.writeBoolean(ech);
        output.writeString(echConfig);

        // version 1
        output.writeString(certPublicKeySha256);

        // version 2
        output.writeString(clientCert);
        output.writeString(clientKey);

        // version 3
        output.writeString(echQueryServerName);
    }

    @Override
    public void deserialize(ByteBufferInput input) {
        int version = input.readInt();
        super.deserialize(input);
        token = input.readString();
        certificates = input.readString();
        udpRelayMode = input.readString();
        congestionController = input.readString();
        alpn = input.readString();
        disableSNI = input.readBoolean();
        zeroRTT = input.readBoolean();
        mtu = input.readInt();
        sni = input.readString();
        allowInsecure = input.readBoolean();
        customJSON = input.readString();
        uuid = input.readString();

        ech = input.readBoolean();
        echConfig = input.readString();

        if (version >= 1) {
            certPublicKeySha256 = input.readString();
        }

        if (version >= 2) {
            clientCert = input.readString();
            clientKey = input.readString();
        }

        if (version >= 3) {
            echQueryServerName = input.readString();
        }
    }

    @Override
    public void applyFeatureSettings(AbstractBean other) {
        if (!(other instanceof TuicBean)) return;
        TuicBean bean = ((TuicBean) other);
    }

    @Override
    public boolean canTCPing() {
        return false;
    }

    @NotNull
    @Override
    public TuicBean clone() {
        return KryoConverters.deserialize(new TuicBean(), KryoConverters.serialize(this));
    }

    @Override
    public @NotNull String outboundType() throws Throwable {
        return SingBoxOptions.TYPE_TUIC;
    }
}

File: app/src/main/kotlin/fr/husi/fmt/tuic/TuicBean.kt
package fr.husi.fmt.tuic

import com.esotericsoftware.kryo.io.ByteBufferInput
import com.esotericsoftware.kryo.io.ByteBufferOutput
import fr.husi.fmt.AbstractBean
import fr.husi.fmt.KryoConverters

class TuicBean : AbstractBean() {

    companion object {
        @JvmField
        val CREATOR = object : CREATOR<TuicBean>() {
            override fun newInstance(): TuicBean {
                return TuicBean()
            }

            override fun newArray(size: Int): Array<TuicBean?> {
                return arrayOfNulls(size)
            }
        }
    }

    var token: String = ""
    var certificates: String = ""
    var certPublicKeySha256: String = ""
    var udpRelayMode: String = "native"
    var congestionController: String = "cubic"
    var alpn: String = ""
    var disableSNI: Boolean = false
    var zeroRTT: Boolean = false
    var mtu: Int = 1400

    // TUIC zep
    var sni: String = ""
    var allowInsecure: Boolean = false
    var customJSON: String = ""
    var uuid: String = ""

    var ech: Boolean = false
    var echConfig: String = ""
    var echQueryServerName: String = ""

    // mTLS
    var clientCert: String = ""
    var clientKey: String = ""

    override fun initializeDefaultValues() {
        super.initializeDefaultValues()
        if (udpRelayMode.isEmpty()) udpRelayMode = "native"
        if (congestionController.isEmpty()) congestionController = "bbr"
    }

    override fun serialize(output: ByteBufferOutput) {
        output.writeInt(3)

        // version 0
        super.serialize(output)
        output.writeString(token)
        output.writeString(certificates)
        output.writeString(udpRelayMode)
        output.writeString(congestionController)
        output.writeString(alpn)
        output.writeBoolean(disableSNI)
        output.writeBoolean(zeroRTT)
        output.writeInt(mtu)
        output.writeString(sni)
        output.writeBoolean(allowInsecure)
        output.writeString(customJSON)
        output.writeString(uuid)
        output.writeBoolean(ech)
        output.writeString(echConfig)

        // version 1
        output.writeString(certPublicKeySha256)

        // version 2
        output.writeString(clientCert)
        output.writeString(clientKey)

        // version 3
        output.writeString(echQueryServerName)
    }

    override fun deserialize(input: ByteBufferInput) {
        val version = input.readInt()
        super.deserialize(input)
        token = input.readString()
        certificates = input.readString()
        udpRelayMode = input.readString()
        congestionController = input.readString()
        alpn = input.readString()
        disableSNI = input.readBoolean()
        zeroRTT = input.readBoolean()
        mtu = input.readInt()
        sni = input.readString()
        allowInsecure = input.readBoolean()
        customJSON = input.readString()
        uuid = input.readString()

        ech = input.readBoolean()
        echConfig = input.readString()

        if (version >= 1) {
            certPublicKeySha256 = input.readString()
        }

        if (version >= 2) {
            clientCert = input.readString()
            clientKey = input.readString()
        }

        if (version >= 3) {
            echQueryServerName = input.readString()
        }
    }

    override fun applyFeatureSettings(other: AbstractBean) {
        if (other !is TuicBean) return
    }

    override val defaultPort get() = 443

    override val canTCPing get() = false

    override fun clone(): TuicBean {
        return KryoConverters.deserialize(TuicBean(), KryoConverters.serialize(this))
    }

}
